# 第15章 離散潜在変数 (Discrete Latent Variables)
## 演習問題 (Exercises 15.1 〜 15.24)

本ノートブックは、Christopher M. Bishop & Hugh Bishop (2024)『深層学習：基礎と概念 (Deep Learning: Foundations and Concepts)』第15章「離散潜在変数」の全演習問題（Exercise 15.1 〜 15.24、全24問）の完全な数学的証明、厳密な導出、および自己採点アサーション付きPython実装を提供します。

---

### 演習問題一覧
1. **Exercise 15.1**: 離散割り当て変数の有限性 ($K^N$ 状態) と歪み尺度 $J$ の狭義単調減少性に基づくK-meansアルゴリズムの有限回収束の証明
2. **Exercise 15.2**: オンライン/逐次型K-meansアルゴリズム（ロビンス・モンロ型確率近似更新式 15.5）の厳密な代数的一致証明
3. **Exercise 15.3**: 1-of-$K$ 表現の潜在変数 $\mathbf{z}$ を持つ結合分布 $p(\mathbf{x}, \mathbf{z})$ からの混合ガウス周辺分布 $p(\mathbf{x}) = \sum_k \pi_k \mathcal{N}(\mathbf{x} \mid \boldsymbol{\mu}_k, \mathbf{\Sigma}_k)$ の導出
4. **Exercise 15.4**: $K$ 成分混合モデルにおける交換対称性（Interchange Symmetries）による等価パラメータ設定数が $K!$ 通り存在することの群論的証明
5. **Exercise 15.5**: 事前分布 $p(\boldsymbol{\theta})$ を持つMAP EMアルゴリズムにおけるEステップの完全同一性とMステップ最大化目的関数 $Q(\boldsymbol{\theta}, \boldsymbol{\theta}^{\text{old}}) + \ln p(\boldsymbol{\theta})$ の導出
6. **Exercise 15.6**: GMMの有向グラフモデル（Figure 15.9）におけるd分離規準を用いた潜在変数の事後分布独立分解 $p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}) = \prod_{n=1}^N p(\mathbf{z}_n \mid \mathbf{x}_n, \boldsymbol{\theta})$ の証明
7. **Exercise 15.7**: 全コンポーネントで共通の共分散行列 $\mathbf{\Sigma}_k = \mathbf{\Sigma}$ を持つ拘束付き混合ガウスモデルに対するEM再推定方程式の導出
8. **Exercise 15.8**: 完全データ対数尤度 $\ln p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta})$ の最大化が各クラスタ独立な標本平均・標本共分散・データ比率に厳密に帰着することの証明
9. **Exercise 15.9**: 不完全データ対数尤度の変分期待値 $Q(\boldsymbol{\theta}, \boldsymbol{\theta}^{\text{old}})$ の $\boldsymbol{\mu}_k$ に関する閉形式停留条件（式 15.16）の導出
10. **Exercise 15.10**: ラグランジュ未定乗数法を用いた $Q$ の $\mathbf{\Sigma}_k$ および $\pi_k$ に関する閉形式極値解（式 15.18, 15.21）の導出
11. **Exercise 15.11**: 任意混合分布の分割変数 $\mathbf{x} = (\mathbf{x}_a, \mathbf{x}_b)$ における条件付き分布 $p(\mathbf{x}_b \mid \mathbf{x}_a)$ が新たな入力依存混合重みを持つ混合分布となることの導出
12. **Exercise 15.12**: 共分散 $\mathbf{\Sigma}_k = \epsilon \mathbf{I}$ の決定論的ゼロ分散極限 $\epsilon \to 0$ において混合ガウスEM法がK-means歪み尺度 $J$ の最小化に厳密に一致することの証明
13. **Exercise 15.13**: 多変量ベルヌーイ分布の期待値 $\mathbb{E}[\mathbf{x}] = \boldsymbol{\mu}$ および対角共分散行列 $\operatorname{cov}[\mathbf{x}] = \operatorname{diag}(\mu_i (1 - \mu_i))$ の導出
14. **Exercise 15.14**: 任意混合分布における全期待値の法則・全分散の法則を用いた平均 $\sum_k \pi_k \boldsymbol{\mu}_k$ および共分散テンソル展開式（式 15.39, 15.40）の証明
15. **Exercise 15.15**: ベルヌーイ混合モデル最尤推定値における標本平均一致性 $\mathbb{E}[\mathbf{x}] = \bar{\mathbf{x}}$ と全成分同一平均初期化時の1反復退化収束の証明
16. **Exercise 15.16**: ベルヌーイ結合分布 $p(\mathbf{x}, \mathbf{z} \mid \boldsymbol{\mu}, \boldsymbol{\pi})$ の $\mathbf{z}$ 周辺化による混合ベルヌーイ分布（式 15.37）の導出
17. **Exercise 15.17**: ベルヌーイ混合モデルの完全データ対数尤度期待値 $Q$ の $\boldsymbol{\mu}_k$ に関するMステップ更新式（式 15.49）の解析的導出
18. **Exercise 15.18**: 単体拘束 $\sum_k \pi_k = 1$ のもとでのラグランジュ乗数法によるベルヌーイ混合係数 $\pi_k = N_k / N$（式 15.50）の導出
19. **Exercise 15.19**: 離散確率質量 $0 \le p(\mathbf{x}_n \mid \boldsymbol{\mu}_k) \le 1$ に起因するベルヌーイ対数尤度の上界（$\ln p(\mathbf{X}) \le 0$）と特異点（発散）不在の証明
20. **Exercise 15.20**: 各次元が $M$ 次のカテゴリカル変数である多変量多項分布混合モデルに対するEステップおよびMステップ再推定方程式の完全導出
21. **Exercise 15.21**: 任意変分分布 $q(\mathbf{Z})$ に対する対数尤度恒等式 $\ln p(\mathbf{X} \mid \boldsymbol{\theta}) = \mathcal{L}(q, \boldsymbol{\theta}) + \mathrm{KL}(q \parallel p)$ の代数的検証
22. **Exercise 15.22**: 変分事後分布 $q(\mathbf{Z}) = p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}^{\text{old}})$ におけるKL勾配消失と下界接線条件 $\nabla_{\boldsymbol{\theta}} \mathcal{L} = \nabla_{\boldsymbol{\theta}} \ln p$ の厳密な証明
23. **Exercise 15.23**: ガウス混合モデルにおける1データ点逐次更新型EM法（Sequential EM）の有効サンプル数およびクラスタ中心更新式（式 15.60, 15.61）の導出
24. **Exercise 15.24**: 逐次更新型EM法における混合係数 $\pi_k$ および共分散行列 $\mathbf{\Sigma}_k$ のオンライン十分統計量増分更新式の完全導出


In [1]:
import sys
import os
from pathlib import Path
import numpy as np
import scipy.stats as stats
import matplotlib.pyplot as plt

# リポジトリルートをパスに追加
repo_root = Path.cwd().parent if Path.cwd().name == "15" else Path.cwd()
if str(repo_root) not in sys.path:
    sys.path.insert(0, str(repo_root))

from common.plot_utils import setup_style
from common.exercises_ch15 import (
    solve_exercise_15_1, verify_exercise_15_1,
    solve_exercise_15_2, verify_exercise_15_2,
    solve_exercise_15_3, verify_exercise_15_3,
    solve_exercise_15_4, verify_exercise_15_4,
    solve_exercise_15_5, verify_exercise_15_5,
    solve_exercise_15_6, verify_exercise_15_6,
    solve_exercise_15_7, verify_exercise_15_7,
    solve_exercise_15_8, verify_exercise_15_8,
    solve_exercise_15_9, verify_exercise_15_9,
    solve_exercise_15_10, verify_exercise_15_10,
    solve_exercise_15_11, verify_exercise_15_11,
    solve_exercise_15_12, verify_exercise_15_12,
    solve_exercise_15_13, verify_exercise_15_13,
    solve_exercise_15_14, verify_exercise_15_14,
    solve_exercise_15_15, verify_exercise_15_15,
    solve_exercise_15_16, verify_exercise_15_16,
    solve_exercise_15_17, verify_exercise_15_17,
    solve_exercise_15_18, verify_exercise_15_18,
    solve_exercise_15_19, verify_exercise_15_19,
    solve_exercise_15_20, verify_exercise_15_20,
    solve_exercise_15_21, verify_exercise_15_21,
    solve_exercise_15_22, verify_exercise_15_22,
    solve_exercise_15_23, verify_exercise_15_23,
    solve_exercise_15_24, verify_exercise_15_24,
    solve_all_exercises,
)

setup_style()
print("Setup complete. Chapter 15 Exercises module successfully loaded.")


Setup complete. Chapter 15 Exercises module successfully loaded.


## Exercises 15.1 〜 15.4: K-meansの収束性と混合モデルの基礎

---

### Exercise 15.1: K-meansアルゴリズムの有限回収束の証明
#### 【問題の要約】
離散割り当て変数 $\{r_{nk}\}$ の可能な組み合わせが有限個（$K^N$ 通り）であり、各割り当てに対してクラスタ代表点 $\{\boldsymbol{\mu}_k\}$ の最適解が一意に定まることから、K-meansアルゴリズムが有限回の反復で必ず収束することを示せ。

#### 【数学的証明】
1. **歪み尺度の定義**:
   $$
   J(\mathbf{R}, \boldsymbol{\mu}) = \sum_{n=1}^N \sum_{k=1}^K r_{nk} \|\mathbf{x}_n - \boldsymbol{\mu}_k\|^2
   $$
   ここで各 $n$ に対して $r_{nk} \in \{0, 1\}$ かつ $\sum_{k=1}^K r_{nk} = 1$ である。割り当て行列 $\mathbf{R} \in \{0, 1\}^{N \times K}$ の可能な配置全体の集合 $\mathcal{S}$ の要素数は有限であり、正確に $|\mathcal{S}| = K^N$ である。
2. **ステップ1（割り当て更新ステップ）**:
   中心 $\{\boldsymbol{\mu}_k\}$ を固定したとき、$r_{nk}$ は各データ点 $\mathbf{x}_n$ を最もユークリッド距離の近い中心に割り当てることで $J$ を最小化する：
   $$
   r_{nk} = \begin{cases} 1 & \text{if } k = \arg\min_j \|\mathbf{x}_n - \boldsymbol{\mu}_j\|^2 \\ 0 & \text{otherwise} \end{cases}
   $$
   したがって、$J(\mathbf{R}^{\text{new}}, \boldsymbol{\mu}) \le J(\mathbf{R}^{\text{old}}, \boldsymbol{\mu})$ である。
3. **ステップ2（代表点更新ステップ）**:
   割り当て行列 $\mathbf{R}$ を固定したとき、$J$ は各中心ベクトル $\boldsymbol{\mu}_k$ に関して狭義凸（2次形式）である。勾配をゼロとおくと：
   $$
   \nabla_{\boldsymbol{\mu}_k} J = 2 \sum_{n=1}^N r_{nk} (\boldsymbol{\mu}_k - \mathbf{x}_n) = \mathbf{0} \implies \boldsymbol{\mu}_k^* = \frac{\sum_{n=1}^N r_{nk} \mathbf{x}_n}{\sum_{n=1}^N r_{nk}}
   $$
   ヘッセ行列は $2 (\sum_n r_{nk}) \mathbf{I} \succ 0$ で正定値であるため、この停留点は**一意な大域的最小値**である。
4. **有限回収束性**:
   各反復サイクルにおいて、$J$ は広義単調減少（$J^{(t+1)} \le J^{(t)}$）する。タイブレークのルールを固定すれば、状態 $\mathbf{R}$ が変化するたびに $J$ は真に減少する。歪み尺度 $J$ は下界 0 を持ち、取り得る状態数 $|\mathcal{S}| = K^N$ は有限であるため、同じ配置を巡回することは不可能であり、アルゴリズムは高々 $K^N$ 回以内の有限反復で必ず完全に静止（収束）する。 $\blacksquare$

---

### Exercise 15.2: 逐次型（オンライン）K-means更新式の導出
#### 【問題の要約】
各ステップで新たなデータ点 $\mathbf{x}_n$ を処理し、最も近い代表点のみを更新する逐次型K-meansを考える。バッチ設定における式 (15.4) から最後のデータ点 $\mathbf{x}_n$ の寄与を分離し、式 (15.5) のロビンス・モンロ型更新式 $\boldsymbol{\mu}_k^{\text{new}} = \boldsymbol{\mu}_k^{\text{old}} + \eta_n (\mathbf{x}_n - \boldsymbol{\mu}_k^{\text{old}})$（ただし $\eta_n = 1/N_k$）となることを示せ。

#### 【数学的導出】
1. **バッチ形式の重心定義**:
   クラスタ $k$ に割り当てられたデータ点数を $N_k$ とすると、式 (15.4) より：
   $$
   \boldsymbol{\mu}_k = \frac{1}{N_k} \sum_{i=1}^N r_{ik} \mathbf{x}_i
   $$
2. **点 $\mathbf{x}_n$ の到着に伴う分離**:
   点 $\mathbf{x}_n$ がクラスタ $k$ に最近接（$r_{nk} = 1$）であるとする。更新前後の割り当て点数は $N_k^{\text{new}} = N_k^{\text{old}} + 1$ である。
   $$
   \boldsymbol{\mu}_k^{\text{new}} = \frac{1}{N_k^{\text{new}}} \left( \sum_{i=1}^{n-1} r_{ik} \mathbf{x}_i + \mathbf{x}_n \right) = \frac{1}{N_k^{\text{new}}} \left( N_k^{\text{old}} \boldsymbol{\mu}_k^{\text{old}} + \mathbf{x}_n \right)
   $$
3. **代数変形**:
   $N_k^{\text{old}} = N_k^{\text{new}} - 1$ を代入すると：
   $$
   \boldsymbol{\mu}_k^{\text{new}} = \frac{N_k^{\text{new}} - 1}{N_k^{\text{new}}} \boldsymbol{\mu}_k^{\text{old}} + \frac{1}{N_k^{\text{new}}} \mathbf{x}_n = \boldsymbol{\mu}_k^{\text{old}} + \frac{1}{N_k^{\text{new}}} (\mathbf{x}_n - \boldsymbol{\mu}_k^{\text{old}})
   $$
4. **学習率 $\eta_n$ との同定**:
   $\eta_n = \frac{1}{N_k^{\text{new}}}$ とおけば、これは確率的勾配降下法・ロビンス・モンロ確率近似アルゴリズムの標準形：
   $$
   \boldsymbol{\mu}_k^{\text{new}} = \boldsymbol{\mu}_k^{\text{old}} + \eta_n (\mathbf{x}_n - \boldsymbol{\mu}_k^{\text{old}}) \quad (\text{式 } 15.5)
   $$
   に厳密に一致する。この導出に近似は一切含まれていないため、逐次代表点は割り当てられた全データの標本平均に常に正確に一致する。 $\blacksquare$

---

### Exercise 15.3: 潜在変数の周辺化による混合ガウス分布の導出
#### 【数学的導出】
潜在変数 $\mathbf{z}$ は 1-of-$K$ 表現のバイナリベクトル（$\mathbf{z} \in \{\mathbf{e}_1, \dots, \mathbf{e}_K\}$）である。
式 (15.9) および式 (15.10) より：
$$
p(\mathbf{z}) = \prod_{k=1}^K \pi_k^{z_k}, \quad p(\mathbf{x} \mid \mathbf{z}) = \prod_{k=1}^K \mathcal{N}(\mathbf{x} \mid \boldsymbol{\mu}_k, \mathbf{\Sigma}_k)^{z_k}
$$
結合分布は乗法定理より：
$$
p(\mathbf{x}, \mathbf{z}) = p(\mathbf{z}) p(\mathbf{x} \mid \mathbf{z}) = \prod_{k=1}^K \left[ \pi_k \mathcal{N}(\mathbf{x} \mid \boldsymbol{\mu}_k, \mathbf{\Sigma}_k) \right]^{z_k}
$$
周辺分布 $p(\mathbf{x})$ は $\mathbf{z}$ の取り得る全 $K$ 個の状態についての和である：
$$
p(\mathbf{x}) = \sum_{\mathbf{z}} p(\mathbf{x}, \mathbf{z}) = \sum_{k=1}^K p(\mathbf{x}, \mathbf{z} = \mathbf{e}_k) = \sum_{k=1}^K \pi_k \mathcal{N}(\mathbf{x} \mid \boldsymbol{\mu}_k, \mathbf{\Sigma}_k) \quad (\text{式 } 15.6) \quad \blacksquare
$$

---

### Exercise 15.4: 混合モデルにおける交換対称性と $K!$ 個の等価モード
#### 【数学的証明】
$K$ 個の成分を持つ混合モデルのパラメータ集合を $\boldsymbol{\theta} = \{(\pi_k, \boldsymbol{\mu}_k, \mathbf{\Sigma}_k)\}_{k=1}^K$ とする。
確率密度関数は $p(\mathbf{x}) = \sum_{k=1}^K \pi_k \mathcal{N}(\mathbf{x} \mid \boldsymbol{\mu}_k, \mathbf{\Sigma}_k)$ である。
足し算の可換性（交換法則）より、インデックス集合 $\{1, \dots, K\}$ の任意の置換 $\sigma \in S_K$（$S_K$ は次数 $K$ の対称群）に対して：
$$
\sum_{k=1}^K \pi_{\sigma(k)} \mathcal{N}(\mathbf{x} \mid \boldsymbol{\mu}_{\sigma(k)}, \mathbf{\Sigma}_{\sigma(k)}) = \sum_{k=1}^K \pi_k \mathcal{N}(\mathbf{x} \mid \boldsymbol{\mu}_k, \mathbf{\Sigma}_k) = p(\mathbf{x})
$$
対称群 $S_K$ の位数（相異なる置換の総数）は $|S_K| = K!$ である。
各成分のパラメータが互いに相異なるとき、これら $K!$ 個の置換はパラメータ空間において相異なる $K!$ 個の点に対応し、いずれも観測データ集合に対して完全に同一の対数尤度値を与える。したがって、交換対称性による等価なパラメータ設定は厳密に $K!$ 通り存在する。 $\blacksquare$


In [2]:
# Verification for Exercises 15.1 - 15.4
print("--- Verifying Exercise 15.1 (K-means finite convergence) ---")
assert verify_exercise_15_1()
sol_15_1 = solve_exercise_15_1()
print(f"Distortion monotonicity and convergence guaranteed in <= {sol_15_1['finite_states']} states.")

print("\n--- Verifying Exercise 15.2 (Sequential K-means update) ---")
assert verify_exercise_15_2()
print("Sequential Robbins-Monro running update matches batch mean with error < 1e-12.")

print("\n--- Verifying Exercise 15.3 (GMM marginalization from p(x, z)) ---")
assert verify_exercise_15_3()
print("Marginalization sum_z p(x, z) matches mixture formula to machine precision.")

print("\n--- Verifying Exercise 15.4 (Interchange symmetries K!) ---")
assert verify_exercise_15_4()
print("All 3! = 6 parameter permutations yield identical log-likelihoods.")


--- Verifying Exercise 15.1 (K-means finite convergence) ---
Distortion monotonicity and convergence guaranteed in <= K^N states.

--- Verifying Exercise 15.2 (Sequential K-means update) ---
Sequential Robbins-Monro running update matches batch mean with error < 1e-12.

--- Verifying Exercise 15.3 (GMM marginalization from p(x, z)) ---
Marginalization sum_z p(x, z) matches mixture formula to machine precision.

--- Verifying Exercise 15.4 (Interchange symmetries K!) ---
All 3! = 6 parameter permutations yield identical log-likelihoods.


## Exercises 15.5 〜 15.8: MAP EM、有向グラフィカルモデルと完全データ尤度

---

### Exercise 15.5: MAP EMアルゴリズムの導出
#### 【問題の要約】
潜在変数を含むモデルにおいて、パラメータの事後分布 $p(\boldsymbol{\theta} \mid \mathbf{X})$ を最大化するMAP EMアルゴリズムを導出せよ。Eステップは最尤推定と同一であり、Mステップでは $Q(\boldsymbol{\theta}, \boldsymbol{\theta}^{\text{old}}) + \ln p(\boldsymbol{\theta})$ を最大化することを示せ。

#### 【数学的導出】
1. **事後分布対数の展開**:
   ベイズの定理より $p(\boldsymbol{\theta} \mid \mathbf{X}) = \frac{p(\mathbf{X} \mid \boldsymbol{\theta}) p(\boldsymbol{\theta})}{p(\mathbf{X})}$ であるから：
   $$
   \ln p(\boldsymbol{\theta} \mid \mathbf{X}) = \ln p(\mathbf{X} \mid \boldsymbol{\theta}) + \ln p(\boldsymbol{\theta}) - \ln p(\mathbf{X})
   $$
   $\ln p(\mathbf{X})$ は $\boldsymbol{\theta}$ に依存しない定数であるため、最大化目標は $\ln p(\mathbf{X} \mid \boldsymbol{\theta}) + \ln p(\boldsymbol{\theta})$ である。
2. **変分下界の導入**:
   第15.4節の分解公式 (15.52) $\ln p(\mathbf{X} \mid \boldsymbol{\theta}) = \mathcal{L}(q, \boldsymbol{\theta}) + \mathrm{KL}(q \parallel p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}))$ を代入すると：
   $$
   \ln p(\boldsymbol{\theta} \mid \mathbf{X}) + \ln p(\mathbf{X}) = \mathcal{L}(q, \boldsymbol{\theta}) + \ln p(\boldsymbol{\theta}) + \mathrm{KL}(q(\mathbf{Z}) \parallel p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}))
   $$
3. **Eステップ**:
   $\boldsymbol{\theta}$ を現在値 $\boldsymbol{\theta}^{\text{old}}$ に固定して $q(\mathbf{Z})$ について最大化する。事前分布対数 $\ln p(\boldsymbol{\theta}^{\text{old}})$ は $q$ に依存しないため、KLダイバージェンスを最小化（0に）する：
   $$
   q^*(\mathbf{Z}) = p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}^{\text{old}})
   $$
   これは通常の最尤推定EMアルゴリズムのEステップと**完全に同一**である。
4. **Mステップ**:
   $q^*(\mathbf{Z}) = p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}^{\text{old}})$ を固定し、目的関数 $\mathcal{L}(q^*, \boldsymbol{\theta}) + \ln p(\boldsymbol{\theta})$ を $\boldsymbol{\theta}$ について最大化する：
   $$
   \mathcal{L}(q^*, \boldsymbol{\theta}) + \ln p(\boldsymbol{\theta}) = \sum_{\mathbf{Z}} q^*(\mathbf{Z}) \ln p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta}) - \sum_{\mathbf{Z}} q^*(\mathbf{Z}) \ln q^*(\mathbf{Z}) + \ln p(\boldsymbol{\theta})
   $$
   第2項は $\boldsymbol{\theta}$ に依存しないエントロピー定数であるから、最大化すべき目的関数は：
   $$
   \mathbb{E}_{q^*}[\ln p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta})] + \ln p(\boldsymbol{\theta}) = Q(\boldsymbol{\theta}, \boldsymbol{\theta}^{\text{old}}) + \ln p(\boldsymbol{\theta}) \quad \blacksquare
   $$

---

### Exercise 15.6: GMM有向グラフのd分離規準による潜在変数事後分布の独立分解
#### 【数学的証明】
Figure 15.9のグラフィカルモデルにおいて、各データ番号 $n \neq m$ について潜在変数 $\mathbf{z}_n$ と $\mathbf{z}_m$ を結ぶ経路を考える。
1. グラフの有向枝はパラメータ $\boldsymbol{\theta} = (\boldsymbol{\mu}, \mathbf{\Sigma}, \boldsymbol{\pi})$ から各 $\mathbf{z}_n$ へ、および $\mathbf{z}_n$ から観測 $\mathbf{x}_n$ へ伸びている（$\boldsymbol{\theta} \to \mathbf{z}_n \to \mathbf{x}_n$）。
2. $\mathbf{z}_n$ と $\mathbf{z}_m$ を結ぶ唯一の経路はパラメータノード $\boldsymbol{\theta}$ を経由する tail-to-tail 構造（$\mathbf{z}_n \leftarrow \boldsymbol{\theta} \rightarrow \mathbf{z}_m$）である。
3. 条件付け集合においてパラメータノード $\boldsymbol{\theta}$ が観測（所与）されているため、この tail-to-tail 経路はブロック（遮断）される。
4. また、観測変数 $\mathbf{x}_n$ は $\mathbf{z}_n$ の子ノードであり、collider（合流点）ではないため、$\mathbf{x}_n$ の観測によって新たな経路が開くことはない。
5. したがって、d分離規準（d-separation criterion）により、$\mathbf{z}_n \perp\!\!\!\perp \mathbf{z}_m \mid (\mathbf{X}, \boldsymbol{\theta})$ が成立する。
6. よって事後分布は各標本ごとに完全に積分解する：
   $$
   p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\mu}, \mathbf{\Sigma}, \boldsymbol{\pi}) = \prod_{n=1}^N p(\mathbf{z}_n \mid \mathbf{x}_n, \boldsymbol{\mu}, \mathbf{\Sigma}, \boldsymbol{\pi}) \quad (\text{式 } 15.62) \quad \blacksquare
   $$

---

### Exercise 15.7: 共通共分散行列 $\mathbf{\Sigma}_k = \mathbf{\Sigma}$ を持つ拘束付きGMMのEM方程式
#### 【数学的導出】
1. **完全データ対数尤度の期待値 $Q$**:
   $$
   Q = \sum_{n=1}^N \sum_{k=1}^K \gamma_{nk} \left[ \ln \pi_k - \frac{D}{2} \ln(2\pi) - \frac{1}{2} \ln |\mathbf{\Sigma}| - \frac{1}{2} (\mathbf{x}_n - \boldsymbol{\mu}_k)^\top \mathbf{\Sigma}^{-1} (\mathbf{x}_n - \boldsymbol{\mu}_k) \right]
   $$
2. **Eステップ**:
   負担率 $\gamma_{nk}$ は共通共分散 $\mathbf{\Sigma}$ を用いて計算される：
   $$
   \gamma_{nk} = \frac{\pi_k \mathcal{N}(\mathbf{x}_n \mid \boldsymbol{\mu}_k, \mathbf{\Sigma})}{\sum_{j=1}^K \pi_j \mathcal{N}(\mathbf{x}_n \mid \boldsymbol{\mu}_j, \mathbf{\Sigma})}
   $$
3. **Mステップ**:
   - 平均 $\boldsymbol{\mu}_k$ および混合重み $\pi_k$ は標準GMMと同一：
     $$
     \boldsymbol{\mu}_k = \frac{1}{N_k} \sum_{n=1}^N \gamma_{nk} \mathbf{x}_n, \quad \pi_k = \frac{N_k}{N}
     $$
   - 共通共分散 $\mathbf{\Sigma}$: 対数行列式項の和は $\sum_n \sum_k \gamma_{nk} = N$ より $-\frac{N}{2} \ln |\mathbf{\Sigma}|$ となる。$\mathbf{\Sigma}^{-1}$ に関する微分をとると：
     $$
     \frac{\partial Q}{\partial \mathbf{\Sigma}^{-1}} = \frac{N}{2} \mathbf{\Sigma} - \frac{1}{2} \sum_{k=1}^K \sum_{n=1}^N \gamma_{nk} (\mathbf{x}_n - \boldsymbol{\mu}_k)(\mathbf{x}_n - \boldsymbol{\mu}_k)^\top = \mathbf{0}
     $$
     したがって：
     $$
     \mathbf{\Sigma} = \frac{1}{N} \sum_{k=1}^K \sum_{n=1}^N \gamma_{nk} (\mathbf{x}_n - \boldsymbol{\mu}_k)(\mathbf{x}_n - \boldsymbol{\mu}_k)^\top = \frac{1}{N} \sum_{k=1}^K N_k \mathbf{S}_k \quad \blacksquare
     $$

---

### Exercise 15.8: 完全データ対数尤度の最大化
#### 【数学的証明】
完全データ対数尤度 (15.26) は以下で与えられる：
$$
\ln p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta}) = \sum_{n=1}^N \sum_{k=1}^K z_{nk} \left[ \ln \pi_k + \ln \mathcal{N}(\mathbf{x}_n \mid \boldsymbol{\mu}_k, \mathbf{\Sigma}_k) \right]
$$
潜在指示変数 $z_{nk} \in \{0, 1\}$ は既知（観測）であるため、データを互いに素なグループ $\mathcal{C}_k = \{n : z_{nk} = 1\}$（サイズ $N_k = \sum_n z_{nk}$）に分割できる：
$$
\ln p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta}) = \sum_{k=1}^K N_k \ln \pi_k + \sum_{k=1}^K \left[ \sum_{n \in \mathcal{C}_k} \ln \mathcal{N}(\mathbf{x}_n \mid \boldsymbol{\mu}_k, \mathbf{\Sigma}_k) \right]
$$
第2項の総和は各クラスタ $k$ ごとに完全に分離している。各クラスタ内の正規分布対数尤度の最大化は単一ガウス分布の最尤推定に帰着し：
$$
\boldsymbol{\mu}_k = \frac{1}{N_k} \sum_{n \in \mathcal{C}_k} \mathbf{x}_n, \quad \mathbf{\Sigma}_k = \frac{1}{N_k} \sum_{n \in \mathcal{C}_k} (\mathbf{x}_n - \boldsymbol{\mu}_k)(\mathbf{x}_n - \boldsymbol{\mu}_k)^\top
$$
また $\pi_k$ については $\sum_k N_k \ln \pi_k$ を $\sum_k \pi_k = 1$ のもとで最大化して $\pi_k = N_k / N$ を得る。 $\blacksquare$


In [3]:
# Verification for Exercises 15.5 - 15.8
print("--- Verifying Exercise 15.5 (MAP EM algorithm) ---")
assert verify_exercise_15_5()
print("MAP EM increases log posterior objective monotonically and stabilizes.")

print("\n--- Verifying Exercise 15.6 (D-separation latent factorization) ---")
assert verify_exercise_15_6()
print("Posterior factorizes p(Z|X, theta) = prod_n p(z_n|x_n, theta) with discrepancy < 1e-14.")

print("\n--- Verifying Exercise 15.7 (Tied covariance GMM EM) ---")
assert verify_exercise_15_7()
print("Tied covariance EM increases log-likelihood monotonically.")

print("\n--- Verifying Exercise 15.8 (Complete-data MLE) ---")
assert verify_exercise_15_8()
print("Complete data MLE coincides exactly with independent cluster sample statistics.")


--- Verifying Exercise 15.5 (MAP EM algorithm) ---
MAP EM increases log posterior objective monotonically and stabilizes.

--- Verifying Exercise 15.6 (D-separation latent factorization) ---
Posterior factorizes p(Z|X, theta) = prod_n p(z_n|x_n, theta) with discrepancy < 1e-14.

--- Verifying Exercise 15.7 (Tied covariance GMM EM) ---
Tied covariance EM increases log-likelihood monotonically.

--- Verifying Exercise 15.8 (Complete-data MLE) ---
Complete data MLE coincides exactly with independent cluster sample statistics.


## Exercises 15.9 〜 15.12: EM最適化解、条件付き混合分布、K-means極限

---

### Exercise 15.9: $Q$ の $\boldsymbol{\mu}_k$ に関する閉形式極値解（式 15.16）の導出
#### 【数学的導出】
式 (15.30) より、$Q(\boldsymbol{\theta}, \boldsymbol{\theta}^{\text{old}})$ の中で $\boldsymbol{\mu}_k$ に依存する項を取り出すと：
$$
-\frac{1}{2} \sum_{n=1}^N \gamma_{nk} (\mathbf{x}_n - \boldsymbol{\mu}_k)^\top \mathbf{\Sigma}_k^{-1} (\mathbf{x}_n - \boldsymbol{\mu}_k)
$$
$\boldsymbol{\mu}_k$ について勾配をとると：
$$
\nabla_{\boldsymbol{\mu}_k} Q = \sum_{n=1}^N \gamma_{nk} \mathbf{\Sigma}_k^{-1} (\mathbf{x}_n - \boldsymbol{\mu}_k) = \mathbf{\Sigma}_k^{-1} \left[ \sum_{n=1}^N \gamma_{nk} \mathbf{x}_n - \boldsymbol{\mu}_k \sum_{n=1}^N \gamma_{nk} \right] = \mathbf{0}
$$
共分散行列 $\mathbf{\Sigma}_k$ は正定値であるため逆行列 $\mathbf{\Sigma}_k^{-1}$ は非特異である。両辺に左から $\mathbf{\Sigma}_k$ を乗じると：
$$
\sum_{n=1}^N \gamma_{nk} \mathbf{x}_n = N_k \boldsymbol{\mu}_k \implies \boldsymbol{\mu}_k = \frac{1}{N_k} \sum_{n=1}^N \gamma_{nk} \mathbf{x}_n \quad (\text{式 } 15.16) \quad \blacksquare
$$

---

### Exercise 15.10: $Q$ の $\mathbf{\Sigma}_k$ および $\pi_k$ に関する閉形式極値解（式 15.18, 15.21）の導出
#### 【数学的導出】
1. **共分散行列 $\mathbf{\Sigma}_k$ の最適化**:
   $Q$ 中で $\mathbf{\Sigma}_k$ に依存する項は：
   $$
   -\frac{1}{2} \sum_{n=1}^N \gamma_{nk} \left[ \ln |\mathbf{\Sigma}_k| + (\mathbf{x}_n - \boldsymbol{\mu}_k)^\top \mathbf{\Sigma}_k^{-1} (\mathbf{x}_n - \boldsymbol{\mu}_k) \right] = -\frac{N_k}{2} \ln |\mathbf{\Sigma}_k| - \frac{1}{2} \operatorname{Tr}\left( \mathbf{\Sigma}_k^{-1} \sum_{n=1}^N \gamma_{nk} (\mathbf{x}_n - \boldsymbol{\mu}_k)(\mathbf{x}_n - \boldsymbol{\mu}_k)^\top \right)
   $$
   $\mathbf{\Sigma}_k^{-1}$ に関する微分公式 $\frac{\partial \ln |\mathbf{\Sigma}|}{\partial \mathbf{\Sigma}^{-1}} = -\mathbf{\Sigma}$ および $\frac{\partial \operatorname{Tr}(\mathbf{\Sigma}^{-1}\mathbf{A})}{\partial \mathbf{\Sigma}^{-1}} = \mathbf{A}$ より：
   $$
   \frac{\partial Q}{\partial \mathbf{\Sigma}_k^{-1}} = \frac{N_k}{2} \mathbf{\Sigma}_k - \frac{1}{2} \sum_{n=1}^N \gamma_{nk} (\mathbf{x}_n - \boldsymbol{\mu}_k)(\mathbf{x}_n - \boldsymbol{\mu}_k)^\top = \mathbf{0}
   $$
   したがって式 (15.18) を得る：
   $$
   \mathbf{\Sigma}_k = \frac{1}{N_k} \sum_{n=1}^N \gamma_{nk} (\mathbf{x}_n - \boldsymbol{\mu}_k)(\mathbf{x}_n - \boldsymbol{\mu}_k)^\top
   $$
2. **混合重み $\pi_k$ のラグランジュ未定乗数法**:
   目的関数は $\sum_{k=1}^K N_k \ln \pi_k$、制約条件は $\sum_{k=1}^K \pi_k = 1$。ラグランジアンは：
   $$
   \Lambda(\boldsymbol{\pi}, \lambda) = \sum_{k=1}^K N_k \ln \pi_k + \lambda \left( \sum_{k=1}^K \pi_k - 1 \right)
   $$
   微分して停留条件を求めると $\frac{N_k}{\pi_k} + \lambda = 0 \implies N_k = -\lambda \pi_k$。
   $k$ について総和をとると $\sum_k N_k = N = -\lambda \sum_k \pi_k = -\lambda \implies \lambda = -N$。
   代入すれば式 (15.21) を得る：
   $$
   \pi_k = \frac{N_k}{N} \quad \blacksquare
   $$

---

### Exercise 15.11: 分割変数における条件付き密度 $p(\mathbf{x}_b \mid \mathbf{x}_a)$ の混合分布構造
#### 【数学的導出】
観測ベクトルを $\mathbf{x} = (\mathbf{x}_a, \mathbf{x}_b)$ と分割する。
周辺密度は：
$$
p(\mathbf{x}_a) = \int p(\mathbf{x}_a, \mathbf{x}_b) d\mathbf{x}_b = \sum_{k=1}^K \pi_k \int p(\mathbf{x}_a, \mathbf{x}_b \mid k) d\mathbf{x}_b = \sum_{k=1}^K \pi_k p(\mathbf{x}_a \mid k)
$$
条件付き密度は乗法定理より：
$$
p(\mathbf{x}_b \mid \mathbf{x}_a) = \frac{p(\mathbf{x}_a, \mathbf{x}_b)}{p(\mathbf{x}_a)} = \frac{\sum_{k=1}^K \pi_k p(\mathbf{x}_a \mid k) p(\mathbf{x}_b \mid \mathbf{x}_a, k)}{\sum_{j=1}^K \pi_j p(\mathbf{x}_a \mid j)}
$$
新たな混合係数 $\widetilde{\pi}_k(\mathbf{x}_a)$ を定義する：
$$
\widetilde{\pi}_k(\mathbf{x}_a) \equiv \frac{\pi_k p(\mathbf{x}_a \mid k)}{\sum_{j=1}^K \pi_j p(\mathbf{x}_a \mid j)} = p(k \mid \mathbf{x}_a)
$$
これは $\widetilde{\pi}_k(\mathbf{x}_a) \ge 0$ かつ $\sum_{k=1}^K \widetilde{\pi}_k(\mathbf{x}_a) = 1$ を満たす正当な確率分布である。
したがって：
$$
p(\mathbf{x}_b \mid \mathbf{x}_a) = \sum_{k=1}^K \widetilde{\pi}_k(\mathbf{x}_a) p(\mathbf{x}_b \mid \mathbf{x}_a, k)
$$
となり、条件付き密度もまた成分分布 $p(\mathbf{x}_b \mid \mathbf{x}_a, k)$ と入力依存の混合重み $\widetilde{\pi}_k(\mathbf{x}_a)$ を持つ**混合分布**となる。 $\blacksquare$

---

### Exercise 15.12: $\epsilon \to 0$ 極限におけるK-means歪み尺度 $J$ との等価性
#### 【数学的証明】
各成分の共分散行列を $\mathbf{\Sigma}_k = \epsilon \mathbf{I}$ とおく。
1. **負担率の極限**:
   $$
   \gamma_{nk} = \frac{\pi_k \exp\left( - \frac{\|\mathbf{x}_n - \boldsymbol{\mu}_k\|^2}{2\epsilon} \right)}{\sum_{j=1}^K \pi_j \exp\left( - \frac{\|\mathbf{x}_n - \boldsymbol{\mu}_j\|^2}{2\epsilon} \right)}
   $$
   $\epsilon \to 0$ の極限では、距離 $\|\mathbf{x}_n - \boldsymbol{\mu}_k\|^2$ が最小となる成分が指数的に圧倒的となり：
   $$
   \lim_{\epsilon \to 0} \gamma_{nk} = r_{nk} \in \{0, 1\} \quad \text{where } r_{nk} = 1 \iff k = \arg\min_j \|\mathbf{x}_n - \boldsymbol{\mu}_j\|^2
   $$
2. **完全データ対数尤度期待値 $Q$ の漸近形**:
   $$
   Q = \sum_{n=1}^N \sum_{k=1}^K \gamma_{nk} \left[ \ln \pi_k - \frac{D}{2} \ln(2\pi\epsilon) - \frac{1}{2\epsilon} \|\mathbf{x}_n - \boldsymbol{\mu}_k\|^2 \right]
   $$
   両辺に $-2\epsilon$ を掛けると：
   $$
   -2\epsilon Q = \sum_{n=1}^N \sum_{k=1}^K r_{nk} \|\mathbf{x}_n - \boldsymbol{\mu}_k\|^2 + \mathcal{O}(\epsilon \ln \epsilon) = J + \mathcal{O}(\epsilon \ln \epsilon)
   $$
   したがって、$\epsilon \to 0$ の極限において $Q$ の最大化はK-meansの歪み尺度 $J$ の最小化に厳密に等価となる。 $\blacksquare$


In [4]:
# Verification for Exercises 15.9 - 15.12
print("--- Verifying Exercise 15.9 (Maximizing Q w.r.t. mu_k) ---")
assert verify_exercise_15_9()
print("Analytic gradient grad_{mu_k} Q vanishes at closed-form solution with norm < 1e-12.")

print("\n--- Verifying Exercise 15.10 (Maximizing Q w.r.t. Sigma_k & pi_k) ---")
assert verify_exercise_15_10()
print("Stationarity conditions and Lagrange multipliers hold identically.")

print("\n--- Verifying Exercise 15.11 (Conditional mixture distribution) ---")
assert verify_exercise_15_11()
print("Conditional density p(x_b|x_a) formula matches joint/marginal ratio to machine precision.")

print("\n--- Verifying Exercise 15.12 (K-means limit epsilon -> 0) ---")
assert verify_exercise_15_12()
print("Soft responsibilities approach hard 0-1 indicator variables as epsilon -> 0.")


--- Verifying Exercise 15.9 (Maximizing Q w.r.t. mu_k) ---
Analytic gradient grad_{mu_k} Q vanishes at closed-form solution with norm < 1e-12.

--- Verifying Exercise 15.10 (Maximizing Q w.r.t. Sigma_k & pi_k) ---
Stationarity conditions and Lagrange multipliers hold identically.

--- Verifying Exercise 15.11 (Conditional mixture distribution) ---
Conditional density p(x_b|x_a) formula matches joint/marginal ratio to machine precision.

--- Verifying Exercise 15.12 (K-means limit epsilon -> 0) ---
Soft responsibilities approach hard 0-1 indicator variables as epsilon -> 0.


## Exercises 15.13 〜 15.16: ベルヌーイ混合モデルと全分散の法則

---

### Exercise 15.13: 多変量ベルヌーイ分布の平均と共分散
#### 【数学的導出】
$p(\mathbf{x} \mid \boldsymbol{\mu}) = \prod_{i=1}^D \mu_i^{x_i} (1 - \mu_i)^{1 - x_i}$（$x_i \in \{0, 1\}$）。
1. 各成分 $x_i$ の期待値：
   $$
   \mathbb{E}[x_i] = 1 \cdot \mu_i + 0 \cdot (1 - \mu_i) = \mu_i \implies \mathbb{E}[\mathbf{x}] = \boldsymbol{\mu} \quad (\text{式 } 15.35)
   $$
2. 2次モーメントおよび分散：
   $$
   \mathbb{E}[x_i^2] = 1^2 \cdot \mu_i + 0^2 \cdot (1 - \mu_i) = \mu_i \implies \operatorname{var}[x_i] = \mathbb{E}[x_i^2] - (\mathbb{E}[x_i])^2 = \mu_i - \mu_i^2 = \mu_i (1 - \mu_i)
   $$
3. 相異なる次元 $i \neq j$ の共分散：
   確率分布が各次元ごとに独立に積分解するため $\mathbb{E}[x_i x_j] = \mathbb{E}[x_i] \mathbb{E}[x_j] = \mu_i \mu_j$。
   したがって $\operatorname{cov}[x_i, x_j] = 0$。
4. 共分散行列全体：
   $$
   \operatorname{cov}[\mathbf{x}] = \operatorname{diag}\left( \mu_1 (1 - \mu_1), \dots, \mu_D (1 - \mu_D) \right) \quad (\text{式 } 15.36) \quad \blacksquare
   $$

---

### Exercise 15.14: 任意混合分布の平均と共分散（全期待値・全分散の法則）
#### 【数学的証明】
1. **平均（全期待値の法則）**:
   $$
   \mathbb{E}[\mathbf{x}] = \mathbb{E}_k[\mathbb{E}[\mathbf{x} \mid k]] = \sum_{k=1}^K \pi_k \boldsymbol{\mu}_k \quad (\text{式 } 15.39)
   $$
2. **2次モーメントテンソル**:
   各成分内の2次モーメントは $\mathbb{E}[\mathbf{x} \mathbf{x}^\top \mid k] = \mathbf{\Sigma}_k + \boldsymbol{\mu}_k \boldsymbol{\mu}_k^\top$ であるから：
   $$
   \mathbb{E}[\mathbf{x} \mathbf{x}^\top] = \sum_{k=1}^K \pi_k \left( \mathbf{\Sigma}_k + \boldsymbol{\mu}_k \boldsymbol{\mu}_k^\top \right)
   $$
3. **共分散行列**:
   $$
   \operatorname{cov}[\mathbf{x}] = \mathbb{E}[\mathbf{x} \mathbf{x}^\top] - \mathbb{E}[\mathbf{x}] \mathbb{E}[\mathbf{x}]^\top = \sum_{k=1}^K \pi_k \left\{ \mathbf{\Sigma}_k + \boldsymbol{\mu}_k \boldsymbol{\mu}_k^\top \right\} - \mathbb{E}[\mathbf{x}] \mathbb{E}[\mathbf{x}]^\top \quad (\text{式 } 15.40)
   $$
   これは全分散の法則 $\operatorname{cov}[\mathbf{x}] = \sum_k \pi_k \mathbf{\Sigma}_k + \sum_k \pi_k (\boldsymbol{\mu}_k - \mathbb{E}[\mathbf{x}])(\boldsymbol{\mu}_k - \mathbb{E}[\mathbf{x}])^\top$ と等価である。 $\blacksquare$

---

### Exercise 15.15: ベルヌーイ混合モデル最尤解の標本平均一致性と1反復退化収束
#### 【数学的証明】
1. **最尤解における期待値の標本平均一致**:
   Mステップの更新式 $\boldsymbol{\mu}_k = \frac{1}{N_k} \sum_n \gamma_{nk} \mathbf{x}_n$ および $\pi_k = \frac{N_k}{N}$ を用いると：
   $$
   \mathbb{E}[\mathbf{x}] = \sum_{k=1}^K \pi_k \boldsymbol{\mu}_k = \sum_{k=1}^K \frac{N_k}{N} \left( \frac{1}{N_k} \sum_{n=1}^N \gamma_{nk} \mathbf{x}_n \right) = \frac{1}{N} \sum_{n=1}^N \left( \sum_{k=1}^K \gamma_{nk} \right) \mathbf{x}_n
   $$
   すべてのデータ点 $n$ について $\sum_{k=1}^K \gamma_{nk} = 1$ であるから：
   $$
   \mathbb{E}[\mathbf{x}] = \frac{1}{N} \sum_{n=1}^N \mathbf{x}_n \equiv \bar{\mathbf{x}} \quad (\text{式 } 15.65)
   $$
2. **同一平均初期化時の1反復退化収束**:
   初期化においてすべての $k$ で $\boldsymbol{\mu}_k = \bar{\boldsymbol{\mu}}$ と設定したとする。
   各データ点 $n$ に対する各成分の条件付き確率は $p(\mathbf{x}_n \mid \bar{\boldsymbol{\mu}})$ で $k$ に無関係に等しくなる。
   したがって負担率は：
   $$
   \gamma_{nk} = \frac{\pi_k p(\mathbf{x}_n \mid \bar{\boldsymbol{\mu}})}{\sum_{j=1}^K \pi_j p(\mathbf{x}_n \mid \bar{\boldsymbol{\mu}})} = \frac{\pi_k}{\sum_j \pi_j} = \pi_k \quad (\text{データ点 } n \text{ に独立})
   $$
   このとき有効点数は $N_k = \sum_{n=1}^N \gamma_{nk} = N \pi_k$ となる。
   直後のMステップにおいて平均を更新すると：
   $$
   \boldsymbol{\mu}_k^{\text{new}} = \frac{1}{N_k} \sum_{n=1}^N \gamma_{nk} \mathbf{x}_n = \frac{1}{N \pi_k} \sum_{n=1}^N \pi_k \mathbf{x}_n = \frac{1}{N} \sum_{n=1}^N \mathbf{x}_n = \bar{\mathbf{x}}
   $$
   すべてのクラスタ中心が直ちに全体の標本平均 $\bar{\mathbf{x}}$ と完全に一致し、以降の反復でも負担率は変化しない。すなわち任意の初期混合係数 $\boldsymbol{\pi}$ に対して**正確に1回の反復で収束**する。 $\blacksquare$

---

### Exercise 15.16: ベルヌーイ結合分布の周辺化
#### 【数学的導出】
式 (15.42) および式 (15.43) より：
$$
p(\mathbf{x} \mid \mathbf{z}, \boldsymbol{\mu}) = \prod_{k=1}^K \left[ \prod_{i=1}^D \mu_{ki}^{x_i} (1 - \mu_{ki})^{1 - x_i} \right]^{z_k}, \quad p(\mathbf{z} \mid \boldsymbol{\pi}) = \prod_{k=1}^K \pi_k^{z_k}
$$
1-of-$K$ ベクトル $\mathbf{z} = \mathbf{e}_k$ のとき $p(\mathbf{x}, \mathbf{e}_k) = \pi_k \prod_{i=1}^D \mu_{ki}^{x_i} (1 - \mu_{ki})^{1 - x_i}$ であるから：
$$
p(\mathbf{x}) = \sum_{\mathbf{z}} p(\mathbf{x}, \mathbf{z}) = \sum_{k=1}^K \pi_k \prod_{i=1}^D \mu_{ki}^{x_i} (1 - \mu_{ki})^{1 - x_i} \quad (\text{式 } 15.37) \quad \blacksquare
$$


In [5]:
# Verification for Exercises 15.13 - 15.16
print("--- Verifying Exercise 15.13 (Multivariate Bernoulli mean & covariance) ---")
assert verify_exercise_15_13()
print("Sample mean and diagonal covariance match analytical formula with error < 0.01.")

print("\n--- Verifying Exercise 15.14 (Law of total variance for mixtures) ---")
assert verify_exercise_15_14()
print("Mixture mean and covariance tensor formulas match Monte Carlo estimates.")

print("\n--- Verifying Exercise 15.15 (Bernoulli MLE sample mean & 1-step convergence) ---")
assert verify_exercise_15_15()
print("Bernoulli EM with identical initial means converges in exactly 1 step to x_bar.")

print("\n--- Verifying Exercise 15.16 (Marginalizing Bernoulli joint distribution) ---")
assert verify_exercise_15_16()
print("Marginalization sum_z p(x, z) equals Bernoulli mixture formula to machine precision.")


--- Verifying Exercise 15.13 (Multivariate Bernoulli mean & covariance) ---
Sample mean and diagonal covariance match analytical formula with error < 0.01.

--- Verifying Exercise 15.14 (Law of total variance for mixtures) ---
Mixture mean and covariance tensor formulas match Monte Carlo estimates.

--- Verifying Exercise 15.15 (Bernoulli MLE sample mean & 1-step convergence) ---
Bernoulli EM with identical initial means converges in exactly 1 step to x_bar.

--- Verifying Exercise 15.16 (Marginalizing Bernoulli joint distribution) ---
Marginalization sum_z p(x, z) equals Bernoulli mixture formula to machine precision.


## Exercises 15.17 〜 15.20: ベルヌーイMステップ、特異点不在、多項混合モデル

---

### Exercise 15.17: ベルヌーイMステップ更新式（式 15.49）の解析的導出
#### 【数学的導出】
ベルヌーイ混合モデルの完全データ対数尤度期待値 (15.45) は：
$$
Q = \sum_{n=1}^N \sum_{k=1}^K \gamma_{nk} \left[ \ln \pi_k + \sum_{i=1}^D \left( x_{ni} \ln \mu_{ki} + (1 - x_{ni}) \ln(1 - \mu_{ki}) \right) \right]
$$
$\mu_{ki}$ について偏微分すると：
$$
\frac{\partial Q}{\partial \mu_{ki}} = \sum_{n=1}^N \gamma_{nk} \left[ \frac{x_{ni}}{\mu_{ki}} - \frac{1 - x_{ni}}{1 - \mu_{ki}} \right] = \sum_{n=1}^N \gamma_{nk} \left[ \frac{x_{ni} - \mu_{ki}}{\mu_{ki}(1 - \mu_{ki})} \right] = 0
$$
$\mu_{ki}(1 - \mu_{ki}) > 0$ であるから、分子の和が 0 となる：
$$
\sum_{n=1}^N \gamma_{nk} (x_{ni} - \mu_{ki}) = 0 \implies \sum_{n=1}^N \gamma_{nk} x_{ni} = \mu_{ki} \sum_{n=1}^N \gamma_{nk} = N_k \mu_{ki}
$$
したがって式 (15.49) を得る：
$$
\mu_{ki} = \frac{1}{N_k} \sum_{n=1}^N \gamma_{nk} x_{ni} \implies \boldsymbol{\mu}_k = \frac{1}{N_k} \sum_{n=1}^N \gamma_{nk} \mathbf{x}_n \quad \blacksquare
$$

---

### Exercise 15.18: ラグランジュ乗数法によるベルヌーイ混合係数 $\pi_k$ の導出
#### 【数学的導出】
$Q$ の中で $\pi_k$ に依存する項は $\sum_{k=1}^K N_k \ln \pi_k$ である。
制約条件 $\sum_k \pi_k = 1$ をラグランジュ乗数 $\lambda$ で組み込んだ関数：
$$
\Lambda(\boldsymbol{\pi}, \lambda) = \sum_{k=1}^K N_k \ln \pi_k + \lambda \left( \sum_{k=1}^K \pi_k - 1 \right)
$$
微分して $\frac{\partial \Lambda}{\partial \pi_k} = \frac{N_k}{\pi_k} + \lambda = 0 \implies N_k = -\lambda \pi_k$。
総和をとると $\sum_k N_k = N = -\lambda \implies \lambda = -N$。
代入すると式 (15.50) を得る：
$$
\pi_k = \frac{N_k}{N} \quad \blacksquare
$$

---

### Exercise 15.19: ベルヌーイ混合モデル対数尤度の上界性と特異点（発散）不在の証明
#### 【数学的証明】
1. 任意のバイナリ観測ベクトル $\mathbf{x}_n \in \{0, 1\}^D$ およびパラメータ $\mu_{ki} \in [0, 1]$ に対し、各次元の確率は $0 \le \mu_{ki}^{x_{ni}} (1 - \mu_{ki})^{1 - x_{ni}} \le 1$ である。
2. 独立積も確率質量であるため：
   $$
   0 \le p(\mathbf{x}_n \mid \boldsymbol{\mu}_k) = \prod_{i=1}^D \mu_{ki}^{x_{ni}} (1 - \mu_{ki})^{1 - x_{ni}} \le 1
   $$
3. 混合分布はこれらの凸結合（$\sum_k \pi_k = 1, \pi_k \ge 0$）であるから：
   $$
   0 \le p(\mathbf{x}_n) = \sum_{k=1}^K \pi_k p(\mathbf{x}_n \mid \boldsymbol{\mu}_k) \le \sum_{k=1}^K \pi_k \cdot 1 = 1
   $$
4. 両辺の自然対数をとると $\ln p(\mathbf{x}_n) \le \ln(1) = 0$ である。
   全データに対する不完全データ対数尤度は：
   $$
   \ln p(\mathbf{X}) = \sum_{n=1}^N \ln p(\mathbf{x}_n) \le 0
   $$
   となり、常に対数尤度は上界 0 で抑えられる。
5. ガウス混合モデルでは、1つの成分の中心がデータ点と一致し分散 $\sigma_k^2 \to 0$ となると確率密度が $+\infty$ に発散する特異点が存在するが、ベルヌーイ混合モデルは確率質量関数（離散分布）であるため、$p(\mathbf{x}_n) \le 1$ が保証され、**尤度関数が無限大に発散する特異点は決して生じない**。 $\blacksquare$

---

### Exercise 15.20: 多変量多項（カテゴリカル）分布混合モデルのEMアルゴリズム
#### 【問題の要約】
$D$ 次元の変数 $\mathbf{x}$ の各次元 $i$ が $M$ 個のカテゴリからなる1-of-$M$ 表現ベクトル $x_{ij} \in \{0, 1\}$（$\sum_{j=1}^M x_{ij} = 1$）であるとき、混合多項分布 (15.66) に対する最尤推定EMアルゴリズムを導出せよ。

#### 【数学的導出】
1. **モデルの定義**:
   $$
   p(\mathbf{x}) = \sum_{k=1}^K \pi_k p(\mathbf{x} \mid \boldsymbol{\mu}_k), \quad p(\mathbf{x} \mid \boldsymbol{\mu}_k) = \prod_{i=1}^D \prod_{j=1}^M \mu_{kij}^{x_{ij}}
   $$
   制約条件は $\sum_{j=1}^M \mu_{kij} = 1$（各 $k, i$ について）および $\sum_{k=1}^K \pi_k = 1$。
2. **Eステップ**:
   各データ点 $n$ に対する事後負担率はベイズの定理より：
   $$
   \gamma_{nk} = \frac{\pi_k p(\mathbf{x}_n \mid \boldsymbol{\mu}_k)}{\sum_{l=1}^K \pi_l p(\mathbf{x}_n \mid \boldsymbol{\mu}_l)} = \frac{\pi_k \prod_{i=1}^D \prod_{j=1}^M \mu_{kij}^{x_{nij}}}{\sum_{l=1}^K \pi_l \prod_{i=1}^D \prod_{j=1}^M \mu_{lij}^{x_{nij}}}
   $$
3. **Mステップ**:
   完全データ対数尤度期待値は：
   $$
   Q = \sum_{n=1}^N \sum_{k=1}^K \gamma_{nk} \left[ \ln \pi_k + \sum_{i=1}^D \sum_{j=1}^M x_{nij} \ln \mu_{kij} \right]
   $$
   - 混合重み $\pi_k$: 制約 $\sum_k \pi_k = 1$ より $\pi_k = \frac{N_k}{N}$（ただし $N_k = \sum_{n=1}^N \gamma_{nk}$）。
   - パラメータ $\mu_{kij}$: 各クラスタ $k$ および各次元 $i$ に対し、制約 $\sum_{j=1}^M \mu_{kij} = 1$ をラグランジュ乗数 $\lambda_{ki}$ で導入する：
     $$
     \Lambda = \sum_{j=1}^M \left( \sum_{n=1}^N \gamma_{nk} x_{nij} \right) \ln \mu_{kij} + \lambda_{ki} \left( \sum_{j=1}^M \mu_{kij} - 1 \right)
     $$
     微分して $\frac{\sum_{n=1}^N \gamma_{nk} x_{nij}}{\mu_{kij}} + \lambda_{ki} = 0 \implies \sum_{n=1}^N \gamma_{nk} x_{nij} = -\lambda_{ki} \mu_{kij}$。
     $j$ について総和をとると、$\sum_{j=1}^M x_{nij} = 1$ より：
     $$
     \sum_{n=1}^N \gamma_{nk} (1) = N_k = -\lambda_{ki} \sum_{j=1}^M \mu_{kij} = -\lambda_{ki} \implies \lambda_{ki} = -N_k
     $$
     したがってMステップ更新式は：
     $$
     \mu_{kij} = \frac{\sum_{n=1}^N \gamma_{nk} x_{nij}}{N_k} \quad \blacksquare
     $$


In [6]:
# Verification for Exercises 15.17 - 15.20
print("--- Verifying Exercise 15.17 (Bernoulli M-step mean update) ---")
assert verify_exercise_15_17()
print("Stationary condition dQ/dmu_ki = 0 satisfied at M-step formula.")

print("\n--- Verifying Exercise 15.18 (Bernoulli M-step pi update) ---")
assert verify_exercise_15_18()
print("Lagrange multiplier conditions hold identically for mixing weights.")

print("\n--- Verifying Exercise 15.19 (Absence of singularities in Bernoulli mixture) ---")
assert verify_exercise_15_19()
print("Bernoulli log-likelihood is rigorously non-positive (<= 0) with no singularities.")

print("\n--- Verifying Exercise 15.20 (Categorical mixture EM algorithm) ---")
assert verify_exercise_15_20()
print("CategoricalMixtureEM monotonically increases log-likelihood and maintains valid distributions.")


--- Verifying Exercise 15.17 (Bernoulli M-step mean update) ---
Stationary condition dQ/dmu_ki = 0 satisfied at M-step formula.

--- Verifying Exercise 15.18 (Bernoulli M-step pi update) ---
Lagrange multiplier conditions hold identically for mixing weights.

--- Verifying Exercise 15.19 (Absence of singularities in Bernoulli mixture) ---
Bernoulli log-likelihood is rigorously non-positive (<= 0) with no singularities.

--- Verifying Exercise 15.20 (Categorical mixture EM algorithm) ---
CategoricalMixtureEM monotonically increases log-likelihood and maintains valid distributions.


## Exercises 15.21 〜 15.24: 変分下界 (ELBO)、接線条件、逐次型EM法

---

### Exercise 15.21: 対数尤度の変分下界とKL分解恒等式の検証
#### 【数学的証明】
式 (15.53) および式 (15.54) より：
$$
\mathcal{L}(q, \boldsymbol{\theta}) = \sum_{\mathbf{Z}} q(\mathbf{Z}) \ln \left\{ \frac{p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta})}{q(\mathbf{Z})} \right\}, \quad \mathrm{KL}(q \parallel p) = - \sum_{\mathbf{Z}} q(\mathbf{Z}) \ln \left\{ \frac{p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta})}{q(\mathbf{Z})} \right\}
$$
両者を加算すると：
$$
\mathcal{L}(q, \boldsymbol{\theta}) + \mathrm{KL}(q \parallel p) = \sum_{\mathbf{Z}} q(\mathbf{Z}) \left[ \ln \frac{p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta})}{q(\mathbf{Z})} - \ln \frac{p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta})}{q(\mathbf{Z})} \right]
$$
対数の差をまとめると分母の $q(\mathbf{Z})$ が完全に相殺される：
$$
= \sum_{\mathbf{Z}} q(\mathbf{Z}) \ln \left\{ \frac{p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta})}{p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta})} \right\}
$$
ベイズの定理（乗法定理）より $\frac{p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta})}{p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta})} = p(\mathbf{X} \mid \boldsymbol{\theta})$ であるから：
$$
= \sum_{\mathbf{Z}} q(\mathbf{Z}) \ln p(\mathbf{X} \mid \boldsymbol{\theta}) = \ln p(\mathbf{X} \mid \boldsymbol{\theta}) \sum_{\mathbf{Z}} q(\mathbf{Z}) = \ln p(\mathbf{X} \mid \boldsymbol{\theta}) \quad (\text{式 } 15.52) \quad \blacksquare
$$

---

### Exercise 15.22: 下界 $\mathcal{L}$ と対数尤度の接線条件（勾配一致）の証明
#### 【数学的証明】
1. 分解恒等式 $\ln p(\mathbf{X} \mid \boldsymbol{\theta}) = \mathcal{L}(q, \boldsymbol{\theta}) + \mathrm{KL}(q \parallel p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}))$ を $\boldsymbol{\theta}$ について微分すると：
   $$
   \nabla_{\boldsymbol{\theta}} \ln p(\mathbf{X} \mid \boldsymbol{\theta}) = \nabla_{\boldsymbol{\theta}} \mathcal{L}(q, \boldsymbol{\theta}) + \nabla_{\boldsymbol{\theta}} \mathrm{KL}(q \parallel p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}))
   $$
2. $q(\mathbf{Z}) = p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}^{\text{old}})$ とおいたとき、KLダイバージェンスは $\boldsymbol{\theta} = \boldsymbol{\theta}^{\text{old}}$ において大域的最小値 0 を達成する。
3. 滑らかな関数の最小値において勾配は必ずゼロベクトルとなるため、$\left. \nabla_{\boldsymbol{\theta}} \mathrm{KL}(q \parallel p(\cdot \mid \mathbf{X}, \boldsymbol{\theta})) \right|_{\boldsymbol{\theta} = \boldsymbol{\theta}^{\text{old}}} = \mathbf{0}$ である。
4. これを解析的に直接確認する：
   $$
   \nabla_{\boldsymbol{\theta}} \mathrm{KL} = - \sum_{\mathbf{Z}} q(\mathbf{Z}) \nabla_{\boldsymbol{\theta}} \ln p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}) = - \sum_{\mathbf{Z}} p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}^{\text{old}}) \left. \frac{\nabla_{\boldsymbol{\theta}} p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta})}{p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta})} \right|_{\boldsymbol{\theta} = \boldsymbol{\theta}^{\text{old}}}
   $$
   $$
   = - \sum_{\mathbf{Z}} \nabla_{\boldsymbol{\theta}} p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}) = - \nabla_{\boldsymbol{\theta}} \left( \sum_{\mathbf{Z}} p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}) \right) = - \nabla_{\boldsymbol{\theta}} (1) = \mathbf{0}
   $$
5. したがって：
   $$
   \left. \nabla_{\boldsymbol{\theta}} \mathcal{L}(q, \boldsymbol{\theta}) \right|_{\boldsymbol{\theta} = \boldsymbol{\theta}^{\text{old}}} = \left. \nabla_{\boldsymbol{\theta}} \ln p(\mathbf{X} \mid \boldsymbol{\theta}) \right|_{\boldsymbol{\theta} = \boldsymbol{\theta}^{\text{old}}} \quad \blacksquare
   $$
   これはFigure 15.16において、下界曲線が現在点 $\boldsymbol{\theta}^{\text{old}}$ において対数尤度曲線と接している（傾きが完全に一致する）幾何学的性質を厳密に裏付けている。

---

### Exercise 15.23: 逐次型EM法における平均と有効点数の増分更新式（式 15.60, 15.61）
#### 【数学的導出】
データ点 $\mathbf{x}_m$ のみの負担率が $\gamma^{\text{old}}(z_{mk})$ から $\gamma^{\text{new}}(z_{mk})$ へ更新されたとする。
1. **有効サンプル数 $N_k$ の更新**:
   $$
   N_k^{\text{new}} = \sum_{n \neq m} \gamma(z_{nk}) + \gamma^{\text{new}}(z_{mk}) = \sum_{n=1}^N \gamma^{\text{old}}(z_{nk}) - \gamma^{\text{old}}(z_{mk}) + \gamma^{\text{new}}(z_{mk})
   $$
   $$
   = N_k^{\text{old}} + \gamma^{\text{new}}(z_{mk}) - \gamma^{\text{old}}(z_{mk}) \quad (\text{式 } 15.61)
   $$
2. **クラスタ中心 $\boldsymbol{\mu}_k$ の更新**:
   十分統計量ベクトルを $\mathbf{s}_k = \sum_{n=1}^N \gamma(z_{nk}) \mathbf{x}_n$ とおく。点 $m$ の更新により：
   $$
   \mathbf{s}_k^{\text{new}} = \mathbf{s}_k^{\text{old}} + \left( \gamma^{\text{new}}(z_{mk}) - \gamma^{\text{old}}(z_{mk}) \right) \mathbf{x}_m
   $$
   定義より $\boldsymbol{\mu}_k^{\text{new}} = \frac{\mathbf{s}_k^{\text{new}}}{N_k^{\text{new}}}$ および $\mathbf{s}_k^{\text{old}} = N_k^{\text{old}} \boldsymbol{\mu}_k^{\text{old}}$ である。
   また式 (15.61) より $N_k^{\text{old}} = N_k^{\text{new}} - (\gamma^{\text{new}} - \gamma^{\text{old}})$ であるから：
   $$
   \boldsymbol{\mu}_k^{\text{new}} = \frac{1}{N_k^{\text{new}}} \left[ \left( N_k^{\text{new}} - (\gamma^{\text{new}} - \gamma^{\text{old}}) \right) \boldsymbol{\mu}_k^{\text{old}} + (\gamma^{\text{new}} - \gamma^{\text{old}}) \mathbf{x}_m \right]
   $$
   $$
   = \boldsymbol{\mu}_k^{\text{old}} + \frac{\gamma^{\text{new}}(z_{mk}) - \gamma^{\text{old}}(z_{mk})}{N_k^{\text{new}}} (\mathbf{x}_m - \boldsymbol{\mu}_k^{\text{old}}) \quad (\text{式 } 15.60) \quad \blacksquare
   $$

---

### Exercise 15.24: 逐次型EM法における混合係数と共分散行列の増分更新式
#### 【数学的導出】
1. **混合重み $\pi_k$ の更新**:
   $$
   \pi_k^{\text{new}} = \frac{N_k^{\text{new}}}{N} = \frac{N_k^{\text{old}} + \gamma^{\text{new}}(z_{mk}) - \gamma^{\text{old}}(z_{mk})}{N} = \pi_k^{\text{old}} + \frac{\gamma^{\text{new}}(z_{mk}) - \gamma^{\text{old}}(z_{mk})}{N}
   $$
2. **共分散行列 $\mathbf{\Sigma}_k$ の更新**:
   非中心2次モーメント十分統計量を $\mathbf{C}_k = \sum_{n=1}^N \gamma(z_{nk}) \mathbf{x}_n \mathbf{x}_n^\top$ とおく。
   点 $m$ の寄与のみを差し替えることで：
   $$
   \mathbf{C}_k^{\text{new}} = \mathbf{C}_k^{\text{old}} + (\gamma^{\text{new}}(z_{mk}) - \gamma^{\text{old}}(z_{mk})) \mathbf{x}_m \mathbf{x}_m^\top
   $$
   $\mathbf{\Sigma}_k = \frac{1}{N_k} \mathbf{C}_k - \boldsymbol{\mu}_k \boldsymbol{\mu}_k^\top$ を用いると：
   $$
   \mathbf{\Sigma}_k^{\text{new}} = \frac{1}{N_k^{\text{new}}} \left[ N_k^{\text{old}} \left( \mathbf{\Sigma}_k^{\text{old}} + \boldsymbol{\mu}_k^{\text{old}} (\boldsymbol{\mu}_k^{\text{old}})^\top \right) + \Delta \gamma_{mk} \mathbf{x}_m \mathbf{x}_m^\top \right] - \boldsymbol{\mu}_k^{\text{new}} (\boldsymbol{\mu}_k^{\text{new}})^\top \quad \blacksquare
   $$


In [7]:
# Verification for Exercises 15.21 - 15.24
print("--- Verifying Exercise 15.21 (ELBO + KL decomposition) ---")
assert verify_exercise_15_21()
print("Identity ln p(X) = ELBO + KL verified to machine precision.")

print("\n--- Verifying Exercise 15.22 (Tangential bound condition) ---")
assert verify_exercise_15_22()
print("Gradients grad_theta ELBO and grad_theta ln p(X) match with relative error < 1e-4.")

print("\n--- Verifying Exercise 15.23 (Sequential EM mean and count updates) ---")
assert verify_exercise_15_23()
print("Incremental mean update (Eq. 15.60 & 15.61) matches batch recomputation to 1e-14.")

print("\n--- Verifying Exercise 15.24 (Sequential EM covariance & mixing coefficient updates) ---")
assert verify_exercise_15_24()
print("Incremental covariance and weight updates match batch recomputation to 1e-12.")


--- Verifying Exercise 15.21 (ELBO + KL decomposition) ---
Identity ln p(X) = ELBO + KL verified to machine precision.

--- Verifying Exercise 15.22 (Tangential bound condition) ---
Gradients grad_theta ELBO and grad_theta ln p(X) match with relative error < 1e-4.

--- Verifying Exercise 15.23 (Sequential EM mean and count updates) ---
Incremental mean update (Eq. 15.60 & 15.61) matches batch recomputation to 1e-14.

--- Verifying Exercise 15.24 (Sequential EM covariance & mixing coefficient updates) ---
Incremental covariance and weight updates match batch recomputation to 1e-12.


## 総合検証 (Summary & Self-Contained Solution Check)

第15章の全24演習問題の解説・厳密証明・数値検証コードが正常に実行され、すべての理論的一致が確認されました。
以下のセルを実行して、全問の解答辞書構造を網羅的に検証します。


In [8]:
# Final comprehensive check across all 24 exercises
all_solutions = solve_all_exercises()
assert len(all_solutions) == 24
for i in range(1, 25):
    key = f"15.{i}"
    assert key in all_solutions, f"Missing solution for Exercise {key}"
    assert "derivation" in all_solutions[key], f"Missing derivation for Exercise {key}"

print(f"All {len(all_solutions)} exercises for Chapter 15 verified successfully!")


All 24 exercises for Chapter 15 verified successfully!
